src/Functions/ztest.cpp

```diff
@@ -0,0 +1,171 @@
+#include <Functions/IFunction.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <boost/math/distributions/normal.hpp>
+
+
+namespace DB
+{
+
+    namespace ErrorCodes
+    {
+        extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    }
+
+
+    class FunctionTwoSampleProportionsZTest : public IFunction
+    {
+    public:
+        static constexpr auto POOLED = "pooled";
+        static constexpr auto UNPOOLED  = "unpooled";
+
+        static constexpr auto name = "proportionsZTest";
+
+        static FunctionPtr create(ContextPtr)
+        {
+            return std::make_shared<FunctionTwoSampleProportionsZTest>();
+        }
+
+        String getName() const override
+        {
+            return name;
+        }
+
+        size_t getNumberOfArguments() const override { return 6; }
+
+        bool useDefaultImplementationForNulls() const override { return false; }
+        bool useDefaultImplementationForConstants() const override { return true; }
+        bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+        static DataTypePtr getReturnType()
+        {
+            DataTypes types
+            {
+                std::make_shared<DataTypeNumber<Float64>>(),
+                std::make_shared<DataTypeNumber<Float64>>(),
+                std::make_shared<DataTypeNumber<Float64>>(),
+                std::make_shared<DataTypeNumber<Float64>>(),
+            };
+
+            Strings names
+            {
+                "z_statistic",
+                "p_value",
+                "confidence_interval_low",
+                "confidence_interval_high"
+            };
+
+            return std::make_shared<DataTypeTuple>(
+                std::move(types),
+                std::move(names)
+            );
+        }
+
+        DataTypePtr getReturnTypeImpl(const DataTypes & /*arguments*/) const override
+        {
+            return getReturnType();
+        }
+
+        static void nan(MutableColumnPtr & to)
+        {
+            const Float64 nan = std::numeric_limits<Float64>::quiet_NaN();
+            Tuple tuple({nan, nan, nan, nan});
+            to->insert(tuple);
+        }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+        {
+            MutableColumnPtr to{getReturnType()->createColumn()};
+            to->reserve(input_rows_count);
+
+            if (!isString(arguments[5].type))
+                throw Exception{"The sixth argument of function " + getName() + " should be String, illegal type: " + arguments[5].type->getName(), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT};
+
+            const ColumnPtr & col_usevar = arguments[5].column;
+
+            for (size_t row_num = 0; row_num < input_rows_count; ++row_num)
+            {
+                const UInt64 successes_x = arguments[0].column->getUInt(row_num);
```

Comment:
This will work extremely slow.
1. Need to avoid virtual calls for getUInt, getFloat64 values from columns.
2. String usevar = col_usevar->getDataAt(row_num).toString(); this probably should be a constant string column, and we need to generate 2 specialization.
3.
```
Tuple tuple({zstat, pvalue, ci_low, ci_high});
to->insert(tuple);
```
We can replace such construction with creation of 4 ColumnFloat64 columns, fill them, and build `Tuple` column only on function return.

Reply:
I fixed it.